dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,205 @@
-// driver. Note that this will not include the allocations that have been made
-// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
-// to make future scheduling decisions.
-type DevicePoolStatus struct {
-	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
-}
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
 
-// AllocatedDevice represents a device that has been allocated from the pool.
-type AllocatedDevice struct {
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	ResourceModel `json:",inline" protobuf:"bytes,4,name=resourceModel"`
 
-	// ClaimUIDs contains the UIDs of the claims to which this device
-	// is allocated.
-	ClaimUIDs []types.UID
+	// To be decided: We could split ResourcePool into spec and status or
+	// extend NamedDevice below so that it reflects the current state of
+	// that device. Like the spec that is data that is coming from a driver,
+	// so adding more information is not a conceptual change. Could be
+	// added in any future release as an extension.
+	//
+	// Much harder is tracking "allocated for". That would be useful to
+	// enable running multiple schedulers for different sets of nodes where
+	// all instances share ownership of some ResourcePool with network-attached
+	// devices. It also would enable a DRA driver's controller to co-exist
+	// with structured parameters in the scheduler.
+	//
+	// If we had transactions in the apiserver, we could combine a status
+	// update of the ResourcePool with a status update of the claim.  But
+	// we don't and adding it [is
+	// hard](https://kubernetes.slack.com/archives/C0EG7JC6T/p1714373064352099).
+	// Without transactions, there will be a risk of leaking resources
+	// and/or races around freeing leaked resources. This would be great to
+	// have in 1.31 because requiring schedulers to use the ResourcePool
+	// status for allocation will be a change of behavior, but it'll be
+	// hard to design and implement in time.
 }
 
-// Device is used to track individual devices in a pool.
-type Device struct {
-	// Name is a driver-specific identifier for the device.
-	// +required
-	Name string `json:"name"`
-
-	// Attributes contain additional metadata that can be used in
-	// constraints. If an attribute name overlaps with the pool attribute,
-	// the device attribute takes precedence.
+// ResourceModel must have one and only one field set.
+type ResourceModel struct {
+	// NamedDevices describes available devices by listing them.
 	//
 	// +optional
-	Attributes []Attribute `json:"attributes,omitempty"`
+	NamedDevices *NamedDevices `json:"namedDevices,omitempty" protobuf:"bytes,1,opt,name=namedResources"`
```

Comment:
> Why not just "Devices"? I don't think the "Named" adds anything.

I kept "NamedDevices" here because this is ultimately where vendors decided how to describe what they have. If we make the name too generic, it becomes odd when adding some other kind of "device" description.

> Another thing I struggle with is having this "big switch" between NamedDevices and "something else" makes the claim API difficult.

Not necessarily. The assumption is that "something else" has to fit into the way how devices are selected and the claim user API overall, so probably they will have attributes.  For example, devices with "consumable resources" would still have some attributes while adding something else for selection.

> At that point I start to wonder if we're just dealing with a different thing, that belongs in its own API.

Yes, if someone needs something that absolute doesn't fit into the API that we define here, then we need something else. But I see quite a few alternatives to the `NamedDevices` that can fit well.

"Consumable resources", if we can't get it merged into "NamedDevices" for 1.31, would be one. A simpler "I have n instances of a device with identical attributes" could be another, but is not be worth adding now (no use cases for it).

Reply:
The tendency is to renamed to "Devices" (https://github.com/kubernetes-sigs/wg-device-management/pull/14/files#r1615731319).